File: src/main/java/cn/edu/jxnu/base/controller/admin/AdminIndexController.java
/* 梦境迷离 (C)2020 */
package cn.edu.jxnu.base.controller.admin;

import cn.edu.jxnu.base.controller.BaseController;
import cn.edu.jxnu.base.entity.BorrowBook;
import cn.edu.jxnu.base.entity.User;
import cn.edu.jxnu.base.service.IBorrowBookService;
import cn.edu.jxnu.base.service.IUserService;
import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder;
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator;
import cn.edu.jxnu.base.utils.Constats;
import cn.edu.jxnu.base.utils.JsonResult;
import java.io.IOException;
import java.util.concurrent.TimeoutException;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.apache.shiro.SecurityUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

/**
 * 主页控制层，公共控制层，不需要权限
 *
 * @author 梦境迷离
 * @version V2.0 2020年11月20日
 */
@Controller
@Slf4j
public class AdminIndexController extends BaseController {

    @Autowired private IUserService userService;

    @Autowired private IBorrowBookService borrowBookService;

    /**
     * 登陆
     *
     * @return String
     */
    @RequestMapping(value = {"/admin/", "/admin/index"})
    public String index() {
        log.info("登录界面");
        return "admin/index";
    }

    /**
     * 首页查询
     *
     * @return String
     */
    @RequestMapping(value = {"/admin/welcome"})
    public String welcome() {
        return "admin/welcome";
    }

    /**
     * 用户个人信息页面
     *
     * @param map map
     * @param response response
     * @param id 用户ID
     * @return String
     */
    @RequestMapping(value = {"/admin/info"})
    public String info(ModelMap map, HttpServletResponse response, Integer id) {
        log.info("用户id: " + id);
        User u =
                (User)
                        SecurityUtils.getSubject()
                                .getSession()
                                .getAttribute(Constats.CURRENTUSER + id);
        if (u != null) {
            map.put("user", u);
        } else {
            map.put("message", "因超时无法获取您的个人信息，即将退出登录");
            redirect(response, "/admin/login");
        }
        return "admin/info";
    }

    /**
     * 用户已借阅书籍页面
     *
     * @return String
     */
    @RequestMapping(value = {"/admin/borrow"})
    public String borrow() {
        // 给页面传入必须的已借书籍
        return "admin/borrow";
    }

    /**
     * 注册，不能拦截
     *
     * @param map map
     * @return String
     */
    @GetMapping(value = {"/assets/regist"})
    public String regist(ModelMap map) {
        // 携带一个map，用于添加注册表单数据
        return "admin/regist/form";
    }

    /**
     * 登录页面的注册专用
     *
     * @param user user
     * @return Mono JsonResult
     */
    @RequestMapping(
            value = {"/assets/edit"},
            method = RequestMethod.POST)
    @ResponseBody
    public Mono<JsonResult> registAdd(User user) {
        try {
            log.info("regist:" + user.toString());
            userService.saveOrUpdate(user).subscribe();
        } catch (Exception e) {
            return Mono.just(JsonResult.failure(e.getMessage()));
        }
        return Mono.just(JsonResult.success("注册成功，3秒后自动回到登录页面"));
    }

    /**
     * 验证用户名【学号】是否已经被注册，委托给用户控制层
     *
     * @param userCode 用户码
     * @param request request
     * @param response response
     * @throws ServletException 异常
     * @throws IOException 异常
     */
    @RequestMapping(value = {"/assets/isAvailable"})
    public void isAvailableUse(
            String userCode, HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        log.info("前台验证账户可用: /assets/isAvailable");
        request.getRequestDispatcher("/admin/user/isAvailable/" + userCode)
                .forward(request, response);
    }

    /**
     * 所有人均可修改的个人信息
     *
     * @param user user
     * @param map map
     * @param response response
     * @return Mono JsonResult
     */
    @ResponseBody
    @RequestMapping(
            value = {"/assets/update"},
            method = RequestMethod.POST)
    public Mono<JsonResult> allCanUpdate(User user, ModelMap map, HttpServletResponse response) {
        log.info("path: /assets/update, user: " + user.toString());
        User u =
                (User)
                        SecurityUtils.getSubject()
                                .getSession()
                                .getAttribute(Constats.CURRENTUSER + user.getId());
        if (u == null) {
            // 已经过期
            // 重定向到登录页面
            map.put("message", "因超时无法获取您的个人信息，即将退出登录");
            redirect(response, "/admin/logout");
        }
        userService.saveOrUpdate(user).subscribe();
        // 更新session
        SecurityUtils.getSubject()
                .getSession()
                .setAttribute(Constats.CURRENTUSER + user.getId(), user);

        return Mono.just(JsonResult.success());
    }

    /**
     * 查询用户已借阅的图书 每个人均可操作，不需要授权。超期不可自主归还
     *
     * @param map map
     * @param uCode 用户码
     * @param request request
     * @param response response
     * @return Mono Page
     */
    @RequestMapping(value = {"/assets/borrowList"})
    @ResponseBody
    public Mono<Page<BorrowBook>> borrowList(
            ModelMap map,
            @RequestParam(value = "uCode") String uCode,
            HttpServletRequest request,
            HttpServletResponse response) {
        Mono<User> uMono = userService.findByUserCode(uCode);
        SimpleSpecificationBuilder<BorrowBook> builder = new SimpleSpecificationBuilder<>();
        String bookName = request.getParameter("inputBookName");
        String bookAuthor = request.getParameter("inputAuthor");
        String bookPress = request.getParameter("inputPublication");
        if (StringUtils.isNotBlank(bookName)) {
            builder.add("bookName", Operator.likeAll.name(), bookName);
        }
        if (StringUtils.isNotBlank(bookAuthor)) {
            builder.add("bookAuthor", Operator.likeAll.name(), bookAuthor);
        }
        if (StringUtils.isNotBlank(bookPress)) {
            builder.add("bookPress", Operator.likeAll.name(), bookPress);
        }
        uMono.map(
                        u ->
                                (User)
                                        SecurityUtils.getSubject()
                                                .getSession()
                                                .getAttribute(Constats.CURRENTUSER + u.getId()))
                .subscribe(
                        user -> {
                            if (user.getId() != null) {
                                builder.add("userId", Operator.eq.name(), user.getId());
                            } else {
                                // 已经过期
                                try {
                                    throw new TimeoutException("因超时无法获取您的个人信息，即将退出登录");
                                } catch (TimeoutException e) {
                                    map.put("message", e.getMessage());
                                } finally {
                                    // 重定向到登录页面
                                    redirect(response, "/admin/logout");
                                }
                            }
                        });
        // 得到已借阅的书籍
        return borrowBookService.findAll(builder.generateSpecification(), getPageRequest(request));
    }
}

File: src/main/java/cn/edu/jxnu/base/controller/admin/AdminIndexController.scala
package cn.edu.jxnu.base.controller.admin

import cn.edu.jxnu.base.controller.BaseController
import org.springframework.beans.factory.annotation.Autowired
import cn.edu.jxnu.base.service.IUserService
import cn.edu.jxnu.base.service.IBorrowBookService
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.ui.Model
import org.springframework.web.servlet.ModelAndView
import org.springframework.stereotype.Controller
import org.springframework.ui.ModelMap
import javax.servlet.http.HttpServletResponse
import lombok.extern.slf4j.Slf4j
import org.apache.shiro.SecurityUtils
import cn.edu.jxnu.base.common.Constats
import java.util.concurrent.TimeoutException
import cn.edu.jxnu.base.entity.User
import cn.edu.jxnu.base.common.JsonResult
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.GetMapping
import javax.servlet.http.HttpServletRequest
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.data.domain.Page
import cn.edu.jxnu.base.entity.BorrowBook
import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder
import org.apache.commons.lang3.StringUtils
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator
import org.springframework.beans.factory.annotation.Required

/**
 * 注意：Scala类的文件名和类名可以不不同
 * @author 梦境迷离.
 * @time 2018年5月15日
 * @version v2.0
 */
@Controller
class AdminIndexController @Autowired() (userService: IUserService, borrowBookService: IBorrowBookService)
    extends BaseController {

    /**
     * 以下没有显示声明返回类型，所以默认返回最后的元素数据类型
     */
    //登陆
    @RequestMapping(value = { Array("/admin/", "/admin/index") })
    def index(model: Model) = {
        "admin/index"
    }

    //首页
    @RequestMapping(value = { Array("/admin/welcome") })
    def welcome() = {
        "admin/welcome"
    }

    //用户个人信息
    @RequestMapping(value = { Array("/admin/info") })
    def info(map: ModelMap, response: HttpServletResponse, @RequestParam(value    = "id", required = false) id: Int) = {

        val subject = SecurityUtils.getSubject();
        print("个人信息页面，rememberMe:" + subject.isRemembered())
        println("用户id:" + id)
        val u = SecurityUtils.getSubject().getSession().getAttribute(Constats.CURRENTUSER + id)
        if (u != null) {
            map.put("user", u)
        } else {
            try {
                throw new TimeoutException("因超时无法获取您的个人信息,即将退出登录")
            } catch {
                case e: Exception => map.put("message", e.getMessage())
            } finally {
                redirect(response, "/admin/login")
            }
        }
        "admin/info"
    }

    //书籍借阅页面
    @RequestMapping(value = { Array("/admin/borrow") })
    def borrow() = {
        "admin/borrow"
    }

    //注册，不能拦截
    @RequestMapping(value = { Array("/assets/regist") })
    def regist(map: ModelMap) = {
        "admin/regist/form"
    }

    //登录页面的注册专用
    @ResponseBody
    @PostMapping(value = { Array("/assets/edit") })
    def registAdd(user: User, map: ModelMap): JsonResult = {
        try {

            println("regist:" + user.toString())
            userService.saveOrUpdate(user)
        } catch {
            case e: Exception => return JsonResult.failure(e.getMessage())
        }
        JsonResult.success("注册成功,3秒后自动回到登录页面")
    }

    //验证用户名【学号】是否已经被注册,委托给用户控制层
    @GetMapping(value = { Array("/assets/isAvailable") })
    def isAvailableUse(userCode: String, request: HttpServletRequest, response: HttpServletResponse) = {
        println("前台验证账户可用")
        request.getRequestDispatcher("/admin/user/isAvailable/" + userCode).forward(request, response)
    }

    //所有人均可修改的个人信息
    @ResponseBody
    @PostMapping(value = { Array("/assets/update") })
    def allCanUpdate(user: User, map: ModelMap): JsonResult = {
        println("user:/assets/update:" + user.toString())
        try {
            val u = SecurityUtils.getSubject().getSession().getAttribute(Constats.CURRENTUSER + user.getId())
            if (u == null) {
                // 已经过期
                try {
                    throw new TimeoutException("因超时无法获取您的个人信息,即将退出登录")
                } catch {
                    case e: TimeoutException => map.put("message", e.getMessage())
                } finally {
                    // 重定向到登录页面
                    Thread.sleep(1000)
                    redirect(response, "/admin/logout")
                }
            }
            userService.saveOrUpdate(user)
            // 更新session
            SecurityUtils.getSubject().getSession().setAttribute(Constats.CURRENTUSER + user.getId(), user)
        } catch {
            case e: Exception => return JsonResult.failure(e.getMessage())
        }

        JsonResult.success("修改成功，下次登陆生效")
    }

    //查询用户已借阅的图书 每个人均可操作，不需要授权。超期不可自主归还
    @ResponseBody
    @RequestMapping(value = { Array("/assets/borrowList") })
    def borrowList(map: ModelMap, @RequestParam(value = "uCode") uCode: String): Page[BorrowBook] = {
        val u = userService.findByUserCode(uCode)
        val builder = new SimpleSpecificationBuilder[BorrowBook]()
        val bookName = request.getParameter("inputBookName")
        val bookAuthor = request.getParameter("inputAuthor")
        val bookPress = request.getParameter("inputPublication")
        if (StringUtils.isNotBlank(bookName)) {
            builder.add("bookName", Operator.likeAll.name(), bookName)
        }
        if (StringUtils.isNotBlank(bookAuthor)) {
            builder.add("bookAuthor", Operator.likeAll.name(), bookAuthor)

        }
        if (StringUtils.isNotBlank(bookPress)) {
            builder.add("bookPress", Operator.likeAll.name(), bookPress)
        }
        val user: User = SecurityUtils.getSubject().getSession()
            .getAttribute(Constats.CURRENTUSER + u.getId()).asInstanceOf[User]
        if (user.getId() != null) {
            builder.add("userId", Operator.eq.name(), user.getId())
        } else {
            try {
                throw new TimeoutException("因超时无法获取您的个人信息,即将退出登录")
            } catch {
                case e: TimeoutException => map.put("message", e.getMessage())
            } finally {
                redirect(response, "/admin/logout")
            }
        }
        borrowBookService.findAll(builder.generateSpecification(), getPageRequest())

    }
}